internal/icinga2/client.go

```diff
@@ -176,19 +177,64 @@ func (client *Client) buildHostServiceEvent(ctx context.Context, result CheckRes
 }
 
 // buildAcknowledgementEvent from the given fields.
-func (client *Client) buildAcknowledgementEvent(ctx context.Context, host, service, author, comment string) (*event.Event, error) {
+func (client *Client) buildAcknowledgementEvent(ctx context.Context, host, service, author, comment, evType string) (*event.Event, error) {
 	ev, err := client.buildCommonEvent(ctx, host, service)
 	if err != nil {
 		return nil, err
 	}
 
-	ev.Type = event.TypeAcknowledgement
+	switch evType {
+	case event.TypeAcknowledgementSet, event.TypeAcknowledgementCleared:
+	default:
+		return nil, fmt.Errorf("unsupporeted acknowledgement event type %q", evType)
+	}
+
+	ev.Type = evType
 	ev.Username = author
 	ev.Message = comment
 
 	return ev, nil
 }
 
+// buildDowntimeEvent from the given fields.
+func (client *Client) buildDowntimeEvent(ctx context.Context, d Downtime, evType string) (*event.Event, error) {
+	ev, err := client.buildCommonEvent(ctx, d.Host, d.Service)
+	if err != nil {
+		return nil, err
+	}
+
+	switch evType {
+	case event.TypeDowntimeRemoved, event.TypeDowntimeStarted, event.TypeDowntimeEnd:
+	default:
+		return nil, fmt.Errorf("unsupported downtime event type %q", evType)
+	}
+
+	ev.Type = evType
+	if ev.Type == event.TypeDowntimeRemoved && d.RemoveTime.Time().IsZero() {
+		ev.Type = event.TypeDowntimeEnd
+	}
+
+	ev.Username = d.Author
+	ev.Message = d.Comment
+
+	return ev, nil
+}
```

Comment:
These two function could take a bool like `buildFlappingEvent()` instead of `evType`. Then you wouldn't need error handling for that parameter and also the IMHO slightly strange mapping of `TypeDowntimeRemoved` to `TypeDowntimeEnd` would look less strange.